### ISSUE TITLE
Is it possible to create css3 animation using this library ?.

### ISSUE THREAD

**Issue by kopax** [reactjs/react-modal#334]
### Summary:

Is it possible using `react-modal` to have that kind of result: http://xue2han.github.io/react-dynamic-modal/]

It provides animation, unfortunately, the way it is made doesn't work well with [styled-components](https://github.com/styled-components/styled-components) context.

**Comment by diasbruno** [comment]
@kopax yes, you can. you will need to control with `ReactModal_after-open` and `ReactModal_before-close`.
to get the `before close` animation correctly, the time interval (`closeTimeoutMS`) must match with `duration`. see [example/basic/app.css](https://github.com/reactjs/react-modal/blob/master/examples/bootstrap/app.css)

**Comment by kopax** [comment]
Thanks for the short explanation, it makes sens.

Regarding the styled-components part : 

React doesn't allow you to write css in js while styled-components does. You can pass a theme to the context of all your components if you use as a wrapper a `<ThemeProvider theme={theme}>`

This is how my Virtual DOM looks like : 


```jsx
const render = (messages) => {
  ReactDOM.render(
    <Provider store={store}>
      <LanguageProvider messages={messages}>
        <ThemeProvider theme={theme}>
          <Router
            history={history}
            routes={rootRoute}
            onUpdate={hashLinkScroll}
            render={
              applyRouterMiddleware(useScroll())
            }
          />
        </ThemeProvider>
      </LanguageProvider>
    </Provider>,
    document.getElementById('app')
  );
};
```

Problem I have with [react-dynamic-modal](http://xue2han.github.io/react-dynamic-modal/) was that it inject it's modal as a brother of `<Provider>` and not inside `<ThemeProvider>` which provide the context so it will lost it's context and ability to get customized using the `theme` provided.

These are my two requirement : 

1. injected under `<App>` (which is the 1st child of `<Router>` in my virtual dom, and first component child of `<body>` in the dom)
1. being able to apply a modal with and without css3 effect.

@diasbruno, would it be possible to see an example incl css3 effect (2) so I can see if (1) can do it?

**Comment by kopax** [comment]
I am struggling in order to get the same animation as in Bootstrap.

I can't get any animation at all, does anyone has an example of __react-modal__ that involve transitions ?

**Comment by nckblu** [comment]
Here is a fade in animation for the wrapper in scss:

```
.ReactModal__Overlay {
  transition: opacity 500ms ease-in-out;
  opacity: 0;

  &--after-open {
    opacity: 1;
  }

  &--before-close {
    opacity: 0;
  }
}
```

And here is a 'fade in up' animation for the modal itself.

```
.ReactModal__Content {
  transition: all 500ms ease-in-out;
  transform: translateY(30px);
  opacity: 0;

  &--after-open {
    opacity: 1;
    transform: translateY(0);
  }

  &--before-close {
    opacity: 0;
    transform: translateY(30px);
  }
}


```

**Comment by kopax** [comment]
Thanks. That would have been helpful when I was still trying but I've now decided to use a different modal. I think this should be an example in the documentation for the others.

**Comment by nckblu** [comment]
Good point @kopax I'll make a PR at some point this week.

**Comment by Grsmto** [comment]
It seems like `ReactModal__Overlay--after-open` class is added directly on opening. Which makes animation impossible (the styles are added straight away as the element is appended to the DOM already with the class on).
Is it only me having this issue?

**Comment by kadikraman** [comment]
I could get the opacity animation working with styled-components by [injecting the style globally](https://github.com/styled-components/styled-components/blob/master/docs/api.md#injectglobal) and setting the `closeTimeoutMS` property on the Modal.

However, the scale animation won't work, because I'm already overriding the content with custom styles including `transform: 'translate(-50%, -50%)',` to centre it. 🤔

**Comment by nckblu** [comment]
@Grsmto have you tried my example above? The issue you describe can be avoided by using a transition on the class if I understand correctly. If you post your code we can have a look.

**Comment by nckblu** [comment]
@kadikraman, you can apply multiple transforms to include the translation and scale:
http://stackoverflow.com/questions/10765755/how-to-apply-multiple-transforms-in-css

**Comment by kadikraman** [comment]
@nckblu True, but because of the `transform: 'translate(-50%, -50%)',` instead of looking like the modal is "scaling from below", it ends up looking like it's "scaling from the right" if that makes sense

**Comment by nckblu** [comment]
@kadikraman ahh I see, that's annoying, perhaps you could center using absolute positioning way instead?

**Comment by nwlieb** [comment]
@Grsmto I am also having this issue... 'Fixed' for now by setting the overlay to transparent.

**Comment by yakirsobelman** [comment]
@Grsmto  and @nckblu I also gave it a try, this worked for me:

`.ReactModal__Overlay {
    background-color: rebeccapurple;
  opacity: 0;
}

 .ReactModal__Overlay--after-open {
    opacity: 1;
   transition: opacity 200ms ease-in-out;
  }

.ReactModal__Overlay--before-close {
    opacity: 0;
  }`

Notice that the difference from @nckblu code is that the transition property is set on the `--after-open` class.
Also `closeTimeoutMS` should be configured to match the transitions ms as mentioned above.

**Comment by jseminck** [comment]
Would there be any interesting building this into the API? A new prop named `easeTransitionTimeMs` (or something) which sets this CSS?
